File: sdks/typescript/test/price-round.test.ts
/**
 * Rounding a price onto the market's tick grid.
 *
 * The grid is anchored at `priceMinimum`, because that is what the server
 * checks: `(price - priceMinimum) % priceTick`. This subtracted
 * `price % priceTick`, anchoring at zero — right whenever the floor is a
 * multiple of the tick, and wrong for the rest in a way that produces a
 * plausible number rather than an error.
 *
 * Same defect as B20's, in a different copy of the same rule. Mirrors 2212858.
 */

import { test } from "node:test";
import assert from "node:assert/strict";

import { priceRound } from "../src/types.ts";
import type { Market } from "../src/types.ts";

function market(minimum: number, maximum: number, tick: number): Market {
  return {
    id: 11, marketplaceId: 1, name: "Stock", description: null, symbol: "STK",
    privateMarket: false, priceMinimum: minimum, priceMaximum: maximum,
    priceTick: tick, unitMinimum: 1, unitMaximum: 100, unitTick: 1,
  };
}

test("a grid anchored at a multiple of the tick", () => {
  assert.equal(priceRound(market(100, 200, 25), 137), 125);
  assert.equal(priceRound(market(100, 200, 25), 125), 125);
});

test("a grid anchored away from zero", () => {
  // 110/135/160/185 are legal; the old code gave 125 for 137.
  const stock = market(110, 199, 25);
  assert.equal(priceRound(stock, 137), 135);
  assert.equal(priceRound(stock, 199), 185);
  assert.equal(priceRound(stock, 110), 110);
});

test("prices outside the range are clamped", () => {
  const stock = market(110, 199, 25);
  assert.equal(priceRound(stock, 5), 110);
  assert.equal(priceRound(stock, 10_000), 185);
});

test("a fixed price market has one legal price", () => {
  // A tick of zero used to divide by zero and return NaN.
  const fixed = market(150, 150, 0);
  assert.equal(priceRound(fixed, 137), 150);
  assert.equal(priceRound(fixed, 9_999), 150);
});

test("every result sits on the grid", () => {
  const stock = market(110, 199, 25);
  for (let price = 0; price <= 300; price++) {
    const rounded = priceRound(stock, price);
    assert.equal((rounded - stock.priceMinimum) % stock.priceTick, 0,
      `rounding ${price} gave ${rounded}`);
    assert.ok(rounded >= stock.priceMinimum && rounded <= stock.priceMaximum);
  }
});

File: sdks/java/fm-sdk/src/test/java/fm/PriceRoundTest.java
package fm;

import static org.assertj.core.api.Assertions.assertThat;

import org.junit.jupiter.api.Test;

/**
 * Rounding a price onto the market's tick grid.
 *
 * <p>The grid is anchored at {@code priceMinimum}, because that is what the
 * server checks: {@code (price - priceMinimum) % priceTick}. This subtracted
 * {@code price % priceTick} instead, anchoring at zero — right whenever the
 * floor is a multiple of the tick, which is most markets, and wrong in a way
 * that produces a plausible number for the rest.
 *
 * <p>Same defect as B20's, in a different copy of the same rule.
 */
class PriceRoundTest {

    private static Market market(long minimum, long maximum, long tick) {
        return new Market(11L, 1L, "Stock", null, "STK", false,
                minimum, maximum, tick, 1L, 100L, 1L);
    }

    /** A floor on the tick: the old arithmetic happened to agree here. */
    @Test
    void aGridAnchoredAtAMultipleOfTheTick() {
        var stock = market(100, 200, 25);

        assertThat(stock.priceRound(137)).isEqualTo(125L);
        assertThat(stock.priceRound(125)).isEqualTo(125L);
    }

    /**
     * A floor off the tick, where it did not. Legal prices are 110/135/160/185,
     * so 137 rounds to 135 — the old code gave 125, which the server refuses
     * with "price is not on a tic".
     */
    @Test
    void aGridAnchoredAwayFromZero() {
        var stock = market(110, 199, 25);

        assertThat(stock.priceRound(137)).isEqualTo(135L);
        assertThat(stock.priceRound(199)).isEqualTo(185L);
        assertThat(stock.priceRound(110)).isEqualTo(110L);
    }

    @Test
    void pricesOutsideTheRangeAreClamped() {
        var stock = market(110, 199, 25);

        assertThat(stock.priceRound(5)).isEqualTo(110L);
        assertThat(stock.priceRound(10_000)).isEqualTo(185L);
    }

    /** A tick of zero is a fixed dimension. This used to divide by zero. */
    @Test
    void aFixedPriceMarketHasOneLegalPrice() {
        var fixed = market(150, 150, 0);

        assertThat(fixed.priceRound(137)).isEqualTo(150L);
        assertThat(fixed.priceRound(150)).isEqualTo(150L);
        assertThat(fixed.priceRound(9_999)).isEqualTo(150L);
    }

    /** Every result is a price the server would accept. */
    @Test
    void everyResultSitsOnTheGrid() {
        var stock = market(110, 199, 25);

        for (long price = 0; price <= 300; price++) {
            long rounded = stock.priceRound(price);
            assertThat((rounded - stock.priceMinimum()) % stock.priceTick())
                    .as("rounding %d gave %d, which is off the tick", price, rounded)
                    .isZero();
            assertThat(rounded).isBetween(stock.priceMinimum(), stock.priceMaximum());
        }
    }
}
